envoy/config/eds_resources_cache.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "envoy/common/optref.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/endpoint/v3/endpoint.pb.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Config {
+
+// An interface for cached resource removed callback.
+class EdsResourceRemovalCallback {
+public:
+  virtual ~EdsResourceRemovalCallback() = default;
+
+  // Invoked when a cached resource is removed from the cache.
+  virtual void onCachedResourceRemoved(absl::string_view resource_name) PURE;
+};
+
+// Represents an xDS resources cache for EDS resources, and currently supports
+// a single config-source (ADS). The motivation is that clusters that are
+// updated (not added) during a CDS response will be able to use the current EDS
+// configuration, thus avoiding an additional EDS response.
```

Comment:
I'll provide some more context here, and let me know if you think this should go into the comment.

A CDS response will create a cluster in warming mode. If an EDS response will not be delivered in 15 seconds, that cluster will become active with an empty assignment (no endpoints). The series of these PRs will allow Envoy to use the previously cached assignment if no EDS response is received by the time the warming ends.

According to the xDS-protocol, Envoy should use the cached EDS assignment as soon as the CDS response arrives. However, Envoy intentionally doesn't do so, because it supports a case where the cluster's endpoints are substantially changed. For example, when a cluster is changed from a non-TLS cluster to a TLS cluster, the old assignment should not be used, and Envoy should use the new assignment.

This work is essentially allowing the use of cached resources, thus not requiring the server to send an assignment if there was no update, while still allowing a use-case as described above.

Reply:
That makes sense, thanks for the explanation.  I think just an extra note in the comments saying the xDS server isn't required to send an EDS response after a CDS response would help clarify things.

Reply:
Clarified the comment.